### PR TITLE
Refresh and automate updating licenses list

### PR DESCRIPTION
**Pull request by Secrus**
Resolves: python-poetry/poetry#6689

- [ ] Added **tests** for changed code. (no significant changes)
- [ ] Updated **documentation** for changed code. (internal changes)

### CODE REVIEW

**Review by colindean — APPROVED** [review]
LGTM, just about picked this up myself as a Hacktoberfest project

File: src/poetry/core/spdx/license.py [hunk 1]
```diff
@@ -1,15 +1,15 @@
 from __future__ import annotations
 
-from collections import namedtuple
+from typing import NamedTuple
 
 
-class License(namedtuple("License", "id name is_osi_approved is_deprecated")):
```
Comment (neersighted) [inline src/poetry/core/spdx/license.py:6]: I'd consider just making this a frozen dataclass instead.

File: src/poetry/core/spdx/license.py [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
+class License(NamedTuple):
     id: str
     name: str
     is_osi_approved: bool
     is_deprecated: bool
 
-    CLASSIFIER_SUPPORTED = {
```
Comment (neersighted) [inline src/poetry/core/spdx/license.py:12]: Do we need to update this list at all?

File: src/poetry/core/spdx/license.py [hunk 3]
```diff
@@ -74,7 +74,7 @@ class License(namedtuple("License", "id name is_osi_approved is_deprecated")):
         "ZPL-2.1",
     }
 
-    CLASSIFIER_NAMES = {
```
Comment (neersighted) [inline src/poetry/core/spdx/license.py:77]: Ditto

**Review by neersighted — DISMISSED** [review]

File: src/poetry/core/spdx/license.py (hunk 1 above)
Comment (Secrus) [reply src/poetry/core/spdx/license.py:6]: Would it be a zero-cost drop-in though? `NamedTuple` is compatible with standard `tuple`, afaik dataclasses are not. I could make it into a dataclass and go with more refactoring if needed. However, now that I think about it, I wonder if the work here shouldn't be separated into 2 PRs, one that only adds workflow and json update and another with module refactor. WDYT?

**Comment by Secrus** [comment]
@neersighted I have added the event to the workflow and reverted changes (don't worry about the commit history, I will clean it up before merging)

File: src/poetry/core/spdx/license.py (hunk 1 above)
Comment (neersighted) [reply src/poetry/core/spdx/license.py:6]: I'd be fine with splitting this into two.

File: .github/workflows/update_licenses.yml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        if: steps.changes.outputs.count > 0
+        env:
+          GITHUB_TOKEN: ${{ steps.generate_token.outputs.token }}
+        # Only open a PR if the branch is not attached to an existing one
+        run: |
+          PR=$(gh pr list --head $BRANCH_NAME --json number -q '.[0].number')
+          if [ -z $PR ]; then
+            gh pr create \
+            --head $BRANCH_NAME \
+            --title "Automated dependency upgrades" \
```
Comment (mkniewallner) [inline .github/workflows/update_licenses.yml:57]:
```suggestion
            --title "Automated licenses upgrade" \
```

**Review by mkniewallner — COMMENTED** [review]
I'm wondering if there's a valid reason for having the updater in `poetry-core` itself. Wouldn't it be more suited to have it in a dedicated `scripts` directory outside the package?

**Comment by Secrus** [comment]
@neersighted @mkniewallner I reduced this PR to just a workflow, I will be refactoring licensing in another PR.

File: .github/workflows/update_licenses.yml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        id: changes
+        run: |
+           echo "count=$(git status --porcelain=v1 2>/dev/null | wc -l) >> $GITHUB_OUTPUT"
+
+      - name: Commit & push changes
+        # Only push if changes exist
+        if: steps.changes.outputs.count > 0
+        run: |
+          git config user.name poetry-cron-bot
+          git config user.email poetry-cron-bot@github.com
```
Comment (colindean) [inline .github/workflows/update_licenses.yml:35]:
```suggestion
          git config user.name github-actions[bot]
          git config user.email 41898282+github-actions[bot]@users.noreply.github.com
```

This should make it look like @github-actions did it in the commit log. I've seen this used elsewhere and I've used it myself [here](https://github.com/colindean/fillertext/blob/2922a721878871626cb1c1f6ec8fbb449a51e803/.github/workflows/release.yml#L34-L35).

**Review by colindean — DISMISSED** [review]

**Review by eamanu — DISMISSED** [review]
Hi, this PR LGTM. It would great have the license.py updated. Any possibility to merge it?

**Comment by eamanu** [comment]
Hi, @Secrus Are you still interest to continue this PR? :-)

**Review by radoering — COMMENTED** [review]
Since no member with more experience in github actions seems to be available, I'll do my best and probably ask some stupid questions. 😉

File: .github/workflows/update_licenses.yml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+      - name: Commit & push changes
+        # Only push if changes exist
+        if: steps.changes.outputs.count > 0
+        run: |
+          git config user.name github-actions[bot]
+          git config user.email 41898282+github-actions[bot]@users.noreply.github.com
+          git add .
+          git commit -m "Automated licenses list update"
+          git push -f origin ${{ github.ref_name }}:$BRANCH_NAME
```
[the remark this answers is not in the document]
Comment (edgarrmondragon) [reply .github/workflows/update_licenses.yml:38]: I think the push and PR steps could be combined into a single one of you folks use [peter-evans/create-pull-request](https://github.com/peter-evans/create-pull-request)

[the remark this answers is not in the document]
Comment (Secrus) [reply .github/workflows/update_licenses.yml]: Thanks for the suggestion @edgarrmondragon

File: .github/workflows/update_licenses.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      - name: Update licenses list
+        run: |
+          python src/poetry/core/spdx/helpers.py
+
+      - name: Create PR if necessary
+        uses: peter-evans/create-pull-request@v5
+        with:
+          token: ${{ steps.generate_token.outputs.token }}
+          commit-message: "Automated licenses list update"
+          author: "github-actins[bot] <41898282+github-actions[bot]@users.noreply.github.com>"
```
Comment (radoering) [inline .github/workflows/update_licenses.yml:39]:
```suggestion
          author: "github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>"
```

File: .github/workflows/update_licenses.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
+name: "Update licenses"
+
+on:
+  workflow_dispatch:
+  schedule:
+    - cron: "0 0 1 * *" # run once a month
+
+env:
+  BRANCH_NAME: "license-autoupdate"
```
Comment (radoering) [inline .github/workflows/update_licenses.yml:9]: Do we need this? It's just used once. We could insert the branch name there.

File: .github/workflows/update_licenses.yml (hunk 7 above)
Comment (Secrus) [reply .github/workflows/update_licenses.yml]: Done

File: .github/workflows/update_licenses.yml (hunk 8 above)
Comment (Secrus) [reply .github/workflows/update_licenses.yml]: leftover, removed

**Comment by radoering** [comment]
LGTM to me now. Can we test it before merging as proposed in https://github.com/python-poetry/poetry-core/pull/489#issuecomment-1273502115?

**Review by radoering — APPROVED** [review]
Probably too much effort to test before merging so let's try.

**Comment by dimbleby** [comment]
are you in a suitable virtual environment when you call the script?  won't it just fail with 'No module named poetry'?

**Comment by Secrus** [comment]
> are you in a suitable virtual environment when you call the script? won't it just fail with 'No module named poetry'?

of course it will, (and actually did), but we accepted that risk. See #622 for fixes

**Comment by dimbleby** [comment]
> are you in a suitable virtual environment when you call the script? won't it just fail with 'No module named poetry'?

ah never mind, #622, I am too slow

**Pull request merged**